### PR TITLE
[BACK-2489] Offset Period support with Deltas

### PR DESCRIPTION
**Pull request by Roukoswarf**

### CODE REVIEW

File: data/summary/types/bgm.go [hunk 1]
```diff
@@ -138,37 +163,172 @@ func (B *BGMBucketData) CalculateStats(r any, _ *time.Time) (bool, error) {
...
+
+	s.CalculateDelta()
+}
+
+func (s *BGMStats) CalculateDelta() {
+	// We do this as a separate pass through the periods as the amount of tracking required to reverse the iteration
+	// and fill this in during the period calculation would likely nullify any benefits, at least with the current
+	// approach.
+
+	for k := range s.Periods {
```
Comment (tjotala) [inline data/summary/types/bgm.go:230]: Is this missing a delta for `GlucoseManagementIndicator`? Is that meaningful?

Comment (tjotala) [reply data/summary/types/bgm.go:230]: Ugh, nevermind. No such thing for BGM.

File: data/summary/types/test/summary.go [hunk 2]
```diff
@@ -184,49 +231,78 @@ func RandomBGMSummary(userId string) *types.Summary[types.BGMStats, *types.BGMSt
...
+
 	for _, period := range []string{"1d", "7d", "14d", "30d"} {
 		datum.Stats.Periods[period] = &types.BGMPeriod{
+			HasAverageGlucose: test.RandomBool(),
 			AverageGlucose: &types.Glucose{
 				Value: test.RandomFloat64FromRange(1, 30),
 				Units: glucose.MmolL,
 			},
-			HasAverageGlucose: test.RandomBool(),
+			AverageGlucoseDelta: pointer.FromAny(test.RandomFloat64FromRange(1, 30)),
```
Comment (tjotala) [inline data/summary/types/test/summary.go:257]: Ditto here

File: data/summary/types/test/summary.go [hunk 3]
```diff
@@ -61,76 +62,121 @@ func RandomCGMSummary(userId string) *types.Summary[types.CGMStats, *types.CGMSt
...
+			GlucoseManagementIndicator:      pointer.FromAny(test.RandomFloat64FromRange(0, 20)),
+			GlucoseManagementIndicatorDelta: pointer.FromAny(test.RandomFloat64FromRange(0, 20)),
 
+			HasAverageGlucose: test.RandomBool(),
 			AverageGlucose: &types.Glucose{
 				Value: test.RandomFloat64FromRange(1, 30),
 				Units: glucose.MmolL,
 			},
-			HasAverageGlucose: test.RandomBool(),
+			AverageGlucoseDelta: pointer.FromAny(test.RandomFloat64FromRange(0, 20)),
```
[the remark this answers is not in the document]
Comment (Roukoswarf) [reply data/summary/types/test/summary.go:99]: It was moved to the top of this area as it did not match the ordering of the variables that follow.

File: data/summary/types/test/summary.go (hunk 2 above)
Comment (Roukoswarf) [reply data/summary/types/test/summary.go:257]: It was moved to the top of this area as it did not match the ordering of the following variables.

**Review by tjotala — COMMENTED** [review]
Someone else with more current working experience in this repo & Golang should review this. Tagged Todd, but not limited to him.

File: data/summary/types/bgm.go [hunk 4]
```diff
@@ -25,52 +26,67 @@ type BGMBucketData struct {
...
 type BGMPeriods map[string]*BGMPeriod
 
 type BGMStats struct {
-	Periods    BGMPeriods                             `json:"periods" bson:"periods"`
-	Buckets    Buckets[BGMBucketData, *BGMBucketData] `json:"buckets" bson:"buckets"`
-	TotalHours int                                    `json:"totalHours" bson:"totalHours"`
+	Periods       BGMPeriods                             `json:"periods" bson:"periods"`
+	OffsetPeriods BGMPeriods                             `json:"offsetPeriods" bson:"offsetPeriods"`
+	Buckets       Buckets[BGMBucketData, *BGMBucketData] `json:"buckets" bson:"buckets"`
+	//DailyBuckets  Buckets[BGMBucketData, *BGMBucketData] `json:"dailyBuckets" bson:"dailyBuckets"`
```
Comment (lostlevels) [inline data/summary/types/bgm.go:88]: Is this commented out for a reason or can it be deleted?

File: data/summary/types/bgm.go [hunk 5]
```diff
@@ -83,7 +99,9 @@ func (*BGMStats) GetDeviceDataType() string {
 
 func (s *BGMStats) Init() {
 	s.Buckets = make(Buckets[BGMBucketData, *BGMBucketData], 0)
+	//s.DailyBuckets = make(Buckets[BGMBucketData, *BGMBucketData], 0)
```
Comment (lostlevels) [inline data/summary/types/bgm.go:102]: Same here with commented out

File: data/summary/types/cgm.go [hunk 6]
```diff
@@ -36,79 +37,103 @@ type CGMBucketData struct {
...
 type CGMPeriods map[string]*CGMPeriod
 
 type CGMStats struct {
-	Periods    CGMPeriods                             `json:"periods" bson:"periods"`
-	Buckets    Buckets[CGMBucketData, *CGMBucketData] `json:"buckets" bson:"buckets"`
-	TotalHours int                                    `json:"totalHours" bson:"totalHours"`
+	Periods       CGMPeriods                             `json:"periods" bson:"periods"`
+	OffsetPeriods CGMPeriods                             `json:"offsetPeriods" bson:"offsetPeriods"`
+	Buckets       Buckets[CGMBucketData, *CGMBucketData] `json:"buckets" bson:"buckets"`
+	//DailyBuckets  Buckets[CGMBucketData, *CGMBucketData] `json:"dailyBuckets" bson:"buckets"`
```
Comment (lostlevels) [inline data/summary/types/cgm.go:135]: Is this commented out code still needed?

File: data/summary/types/cgm.go [hunk 7]
```diff
@@ -121,7 +146,9 @@ func (*CGMStats) GetDeviceDataType() string {
 
 func (s *CGMStats) Init() {
 	s.Buckets = make(Buckets[CGMBucketData, *CGMBucketData], 0)
+	//s.DailyBuckets = make(Buckets[CGMBucketData, *CGMBucketData], 0)
```
Comment (lostlevels) [inline data/summary/types/cgm.go:149]: same with this comment

**Review by lostlevels — DISMISSED** [review]
Looks fine minus various commented out code. Not sure if they are for the future or should be removed.

**Comment by toddkazakov** [comment]
I agree with @lostlevels. All commented code should be removed.

**Comment by Roukoswarf** [comment]
I will stash the comments in a branch locally so i dont lose them, as they represent a possible but now unlikely path forward and remove them from the PR.

**Review by lostlevels — DISMISSED** [review]

**Review by toddkazakov — COMMENTED** [review]
The changes to the summary structure are not backward compatible with clinic-worker (thus not compatible with the EHR integration). Please make sure you have corresponding PRs that are ready to be merged before merging this one.

File: data/summary/types/bgm.go [hunk 8]
```diff
@@ -138,37 +161,173 @@ func (B *BGMBucketData) CalculateStats(r any, _ *time.Time) (bool, error) {
...
+		}
+
+		if s.Periods[k].TimeInTargetPercent != nil && s.OffsetPeriods[k].TimeInTargetPercent != nil {
+			delta := *s.Periods[k].TimeInTargetPercent - *s.OffsetPeriods[k].TimeInTargetPercent
+
+			s.Periods[k].TimeInTargetPercentDelta = pointer.FromAny(delta)
+			s.OffsetPeriods[k].TimeInTargetPercentDelta = pointer.FromAny(-delta)
+		}
+
+		if s.Periods[k].TimeInTargetPercent != nil && s.OffsetPeriods[k].TimeInTargetPercent != nil {
```
[the remark this answers is not in the document]
Comment (Roukoswarf) [reply data/summary/types/bgm.go:258]: yes, copy paste error, fixed it in the CGM stats, but missed it here, thank you.

**Review by toddkazakov — APPROVED** [review]

**Pull request merged**